File: solutions/104. Maximum Depth of Binary Tree /104.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        q = deque()
        q.append(root)
        depth = 0
        while q:
            sz = len(q)
            for i in range(sz):
                temp = q.popleft()
                if temp.left:
                    q.append(temp.left)
                if temp.right:
                    q.append(temp.right)
            depth += 1
        return depth

File: neetcode/104. Maximum Depth of Binary Tree/104.kt
/*
Method 1.0:
Intution: Base condition: root == null, depth = 0. Get the max of heights of left and right
subtrees and add 1 (root node)
TC: O(N), where N is no. of nodes, all nodes are traversed
SC: O(log N), recursive calls stack size
 */
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun maxDepth(root: TreeNode?): Int {
        if (root == null) return 0

        return max(maxDepth(root.left), maxDepth(root.right)) + 1
    }
}
